pkg/operator/controller/dns/controller.go

```diff
@@ -0,0 +1,167 @@
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	runtimecontroller "sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "dns_controller"
+)
+
+var log = logf.Logger.WithName(controllerName)
+
+func New(mgr manager.Manager, operatorNamespace string, dnsProvider dns.Provider) (runtimecontroller.Controller, error) {
+	reconciler := &reconciler{
+		client:            mgr.GetClient(),
+		cache:             mgr.GetCache(),
+		recorder:          mgr.GetEventRecorderFor(controllerName),
+		operatorNamespace: operatorNamespace,
+		dnsProvider:       dnsProvider,
+	}
+	c, err := runtimecontroller.New(controllerName, mgr, runtimecontroller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &iov1.DNSRecord{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+type reconciler struct {
+	client            client.Client
+	cache             cache.Cache
+	recorder          record.EventRecorder
+	operatorNamespace string
+	dnsProvider       dns.Provider
+	dnsConfig         *configv1.DNS
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// TODO: If the new revision has changed, undefined stuff happens (resource leaks?)
+	dnsConfig := &configv1.DNS{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, dnsConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get dns 'cluster': %v", err)
+	}
+
+	record := &iov1.DNSRecord{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, record); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("dnsrecord not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		} else {
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, fmt.Errorf("failed to get dnsrecord: %v", err)
+		}
+	}
+
+	// If the DNS record was deleted, clean up and return.
+	if record.DeletionTimestamp != nil {
+		if err := r.delete(record); err != nil {
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	ingressName, ok := record.Labels[manifests.OwningIngressControllerLabel]
+	if !ok {
+		return reconcile.Result{}, fmt.Errorf("missing owner label")
+	}
+	ingress := &operatorv1.IngressController{}
+	if err := r.cache.Get(context.TODO(), types.NamespacedName{Namespace: record.Namespace, Name: ingressName}, ingress); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get dnsrecord for record %s: %v", record.Name, err)
```

Comment:
Is it a deliberate design choice not to handle orphaned `DNSRecord` resources?

Reply:
I don't know that this state implies the `dnsrecord` itself an orphan. It usually means the resource was deleted and due to timing we have a benign leftover entry to flush from the controller queue.

Also, how would the resource have been deleted from storage without having passed through finalization in `delete()`? So I'm also not sure there's any orphan in the cloud provider, either.

Help me understand the concern?

Reply:
There is a possible state that we need to consider: a `DNSRecord` exists with no corresponding `IngressController`.  If you want to say that we will do nothing to try to recover from this state because the only ways to get there involve, say, etcd corruption or user tinkering, then that is fine, but I want to make sure we don't simply ignore the case.

Reply:
Okay, so I tried to split the difference for now. I handle not found by logging a specific message, added a TODO, and made the retry far less aggressive in this case. For other errors, I retry somewhat less aggressively since those are likely transient.